## Final Project Submission

Please fill out:
* Student name: GRACE KOKI MUTUKU
* Student pace: FULL TIME
* Scheduled project review date/time: 
* Instructor name: 
* Blog post URL:


In [1]:
# Import the necessary libraries
import pandas as pd
import numpy as np
import sqlite3
import zipfile
import os

path = r"D:\MORINGA\MoringaGit\dsc-data-science-env-config\Phase_1\Week 3\dsc-phase-1-project-v2-4\zippedData"

In [2]:
os.chdir(path)
os.listdir()

['bom.movie_gross.csv.gz',
 'im.db',
 'im.db.zip',
 'rt.movie_info.tsv.gz',
 'rt.reviews.tsv.gz',
 'tmdb.movies.csv.gz',
 'tn.movie_budgets.csv.gz']

In [3]:
with zipfile.ZipFile('im.db.zip') as f:
    f.extractall()    

In [4]:
conn = sqlite3.connect("im.db")
df = pd.read_sql("""SELECT name FROM sqlite_master WHERE type = 'table';""", conn)
df

name
0   movie_basics
1      directors
2      known_for
3     movie_akas
4  movie_ratings
5        persons
6     principals
7        writers

In [306]:
q = """
SELECT *
  FROM writers
;
"""
pd.read_sql(q, conn).head()

movie_id  person_id
0  tt0285252  nm0899854
1  tt0438973  nm0175726
2  tt0438973  nm1802864
3  tt0462036  nm1940585
4  tt0835418  nm0310087

In [309]:
q = """
SELECT *
  FROM persons
;
"""
pd.read_sql(q, conn).head()

person_id       primary_name  birth_year  death_year  \
0  nm0061671  Mary Ellen Bauder         NaN         NaN   
1  nm0061865       Joseph Bauer         NaN         NaN   
2  nm0062070         Bruce Baum         NaN         NaN   
3  nm0062195       Axel Baumann         NaN         NaN   
4  nm0062798        Pete Baxter         NaN         NaN   

                                 primary_profession  
0         miscellaneous,production_manager,producer  
1        composer,music_department,sound_department  
2                        miscellaneous,actor,writer  
3  camera_department,cinematographer,art_department  
4  production_designer,art_department,set_decorator

In [310]:
q = """
SELECT *
  FROM principals
;
"""
pd.read_sql(q, conn).head()

movie_id  ordering  person_id  category       job        characters
0  tt0111414         1  nm0246005     actor      None       ["The Man"]
1  tt0111414         2  nm0398271  director      None              None
2  tt0111414         3  nm3739909  producer  producer              None
3  tt0323808        10  nm0059247    editor      None              None
4  tt0323808         1  nm3579312   actress      None  ["Beth Boothby"]

In [311]:
q = """
SELECT *
  FROM known_for
;
"""
pd.read_sql(q, conn).head()

person_id   movie_id
0  nm0061671  tt0837562
1  nm0061671  tt2398241
2  nm0061671  tt0844471
3  nm0061671  tt0118553
4  nm0061865  tt0896534

In [312]:
q = """
SELECT *
  FROM movie_basics
;
"""
pd.read_sql(q, conn).head()

movie_id                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

In [313]:
q = """
SELECT *
  FROM movie_ratings
;
"""
pd.read_sql(q, conn).head()

movie_id  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

In [314]:
q = """
SELECT *
  FROM movie_akas
;
"""
pd.read_sql(q, conn).head()

movie_id  ordering                                    title region  \
0  tt0369610        10                            Джурасик свят     BG   
1  tt0369610        11                        Jurashikku warudo     JP   
2  tt0369610        12  Jurassic World: O Mundo dos Dinossauros     BR   
3  tt0369610        13                  O Mundo dos Dinossauros     BR   
4  tt0369610        14                           Jurassic World     FR   

  language        types   attributes  is_original_title  
0       bg         None         None                0.0  
1     None  imdbDisplay         None                0.0  
2     None  imdbDisplay         None                0.0  
3     None         None  short title                0.0  
4     None  imdbDisplay         None                0.0

## Importing Data

The `read_csv()` function is used to import all the csv files into Pandas DataFrame then the proper encoding of `latin-1` is applied.

In [252]:
# Import all the '.csv' file using a proper encoding
movies_df = pd.read_csv('tmdb.movies.csv.gz', encoding='latin-1', index_col= 0)
movie_gross_df = pd.read_csv('bom.movie_gross.csv.gz', encoding='latin-1')
movie_budgets_df = pd.read_csv('tn.movie_budgets.csv.gz', encoding='latin-1', index_col= 0)
movie_info_df = pd.read_csv('rt.movie_info.tsv.gz', delimiter='\t', encoding='latin-1')
reviews_df = pd.read_csv('rt.reviews.tsv.gz', delimiter= '\t', encoding='latin-1')

In [7]:
#movies_df 
# movie_gross_df
# movie_budgets_df
# movie_info_df 
# reviews_df
# movies_df['vote_count'].mode()

In [97]:
def dataframe_structure(df):
    print(f'The shape of the dataframe is: {df.shape}\n')
    df.info()
    display(df.describe())
    print(f'Checking for Null Values: \n',(df.isna().sum()))
    display(df.head())

In [9]:
dataframe_structure(movies_df)

The shape of the dataframe is: (26517, 9)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26517 entries, 0 to 26516
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   genre_ids          26517 non-null  object 
 1   id                 26517 non-null  int64  
 2   original_language  26517 non-null  object 
 3   original_title     26517 non-null  object 
 4   popularity         26517 non-null  float64
 5   release_date       26517 non-null  object 
 6   title              26517 non-null  object 
 7   vote_average       26517 non-null  float64
 8   vote_count         26517 non-null  int64  
dtypes: float64(2), int64(2), object(5)
memory usage: 2.0+ MB


id    popularity  vote_average    vote_count
count   26517.000000  26517.000000  26517.000000  26517.000000
mean   295050.153260      3.130912      5.991281    194.224837
std    153661.615648      4.355229      1.852946    960.961095
min        27.000000      0.600000      0.000000      1.000000
25%    157851.000000      0.600000      5.000000      2.000000
50%    309581.000000      1.374000      6.000000      5.000000
75%    419542.000000      3.694000      7.000000     28.000000
max    608444.000000     80.773000     10.000000  22186.000000

genre_ids     id original_language  \
0      [12, 14, 10751]  12444                en   
1  [14, 12, 16, 10751]  10191                en   
2        [12, 28, 878]  10138                en   
3      [16, 35, 10751]    862                en   
4        [28, 878, 12]  27205                en   

                                 original_title  popularity release_date  \
0  Harry Potter and the Deathly Hallows: Part 1      33.533   2010-11-19   
1                      How to Train Your Dragon      28.734   2010-03-26   
2                                    Iron Man 2      28.515   2010-05-07   
3                                     Toy Story      28.005   1995-11-22   
4                                     Inception      27.920   2010-07-16   

                                          title  vote_average  vote_count  
0  Harry Potter and the Deathly Hallows: Part 1           7.7       10788  
1                      How to Train Your Dragon           7.7        7610  
2                                    Iron Man 2           6.8       12368  
3                                     Toy Story           7.9       10174  
4                                     Inception           8.3       22186

In [10]:
movies_df.isna().sum()

genre_ids            0
id                   0
original_language    0
original_title       0
popularity           0
release_date         0
title                0
vote_average         0
vote_count           0
dtype: int64

There seems to be no missing values in the `tmdb.movies.csv.gz` file since there are 26517 entries in all columns and also checking for the sum of `NaN` values returs zero in every column confirming the absence of null values.

The next step will be to check for the presence of duplicate entries in the `movie_df` based on the subset `id` which is assumed to be unique.

In [11]:
duplicates = movies_df[movies_df.duplicated(subset = 'id')] # Identify duplicate rows based on id
movies_df_no_duplicates = movies_df.drop_duplicates() # Drop the duplicate rows
movies_df_no_duplicates.head()

genre_ids     id original_language  \
0      [12, 14, 10751]  12444                en   
1  [14, 12, 16, 10751]  10191                en   
2        [12, 28, 878]  10138                en   
3      [16, 35, 10751]    862                en   
4        [28, 878, 12]  27205                en   

                                 original_title  popularity release_date  \
0  Harry Potter and the Deathly Hallows: Part 1      33.533   2010-11-19   
1                      How to Train Your Dragon      28.734   2010-03-26   
2                                    Iron Man 2      28.515   2010-05-07   
3                                     Toy Story      28.005   1995-11-22   
4                                     Inception      27.920   2010-07-16   

                                          title  vote_average  vote_count  
0  Harry Potter and the Deathly Hallows: Part 1           7.7       10788  
1                      How to Train Your Dragon           7.7        7610  
2                                    Iron Man 2           6.8       12368  
3                                     Toy Story           7.9       10174  
4                                     Inception           8.3       22186

In [12]:
for column in movies_df.columns:
    print(column, '\n', movies_df_no_duplicates[column].value_counts(normalize = True), '\n\n')

genre_ids 
 [99]                             0.139820
[]                               0.096521
[18]                             0.083108
[35]                             0.063615
[27]                             0.044123
                                   ...   
[27, 53, 99, 18]                 0.000039
[53, 9648, 12, 10749, 27, 14]    0.000039
[878, 18, 28, 10770]             0.000039
[28, 36, 10752]                  0.000039
[35, 10770, 10402]               0.000039
Name: genre_ids, Length: 2477, dtype: float64 


id 
 71677     0.000039
478883    0.000039
105945    0.000039
298459    0.000039
414455    0.000039
            ...   
302150    0.000039
234567    0.000039
451657    0.000039
64586     0.000039
524288    0.000039
Name: id, Length: 25497, dtype: float64 


original_language 
 en    0.877907
fr    0.018943
es    0.017218
ru    0.011570
ja    0.009570
        ...   
sn    0.000039
mi    0.000039
ha    0.000039
dz    0.000039
hy    0.000039
Name: original_language, Length: 76

In [13]:
# movies_df_no_duplicates.loc[movies_df_no_duplicates['genre_ids'] == '[]', :]
# movies_df[movies_df.duplicated(subset = 'title')]

In [14]:
# movies_df_no_duplicates.loc[movies_df_no_duplicates['title'] == 'Eden', : ]

In [15]:
dataframe_structure(movie_gross_df)

The shape of the dataframe is: (3387, 5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB


domestic_gross         year
count    3.359000e+03  3387.000000
mean     2.874585e+07  2013.958075
std      6.698250e+07     2.478141
min      1.000000e+02  2010.000000
25%      1.200000e+05  2012.000000
50%      1.400000e+06  2014.000000
75%      2.790000e+07  2016.000000
max      9.367000e+08  2018.000000

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010

In [16]:
# movie_gross_df.tail(800)

## Understanding the data
There are 5 columns and 3387 entries in the `bom.movie_gross.csv` file. `studio`, `domestic_gross` and `foreign_gross` have missing values. There seems to be no `NaN` values in the file.

The columns have appropriate datatypes apart from `foreign_gross` which should be a `float` value and `year` which should be type `date`.

In [17]:
movie_gross_df['foreign_gross'] = movie_gross_df['foreign_gross'].str.replace(',', '') # Replace the ',' with an empty string, so as to be able to convert to float
movie_gross_df['foreign_gross'] = movie_gross_df['foreign_gross'].astype(float) # Convert the 'foreign_gross' column datatype from object to a float

In [18]:
movie_gross_df['year'] = pd.to_datetime(movie_gross_df['year'], format = '%Y') # Convert the datatype of 'year' from int to datetime


In [58]:
movie_gross_df.describe().round(1)

domestic_gross
count          3359.0
mean       28745845.1
std        66982498.2
min             100.0
25%          120000.0
50%         1400000.0
75%        27900000.0
max       936700000.0

## Cleaning of the data
1. Check for duplicates

In [20]:
duplicates = movie_gross_df[movie_gross_df.duplicated()] # Checking for duplicate entries
len(duplicates)

0

There are no duplicate rows.

In [214]:
movie_gross_df.isna().sum()

title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64

Since 39% of the values in `foreign_gross` are null, it is preferable to drop the entire column rather than replace the missing values or drop the rows with missing values and distort data of loose invaluable data respectively. 

In [25]:
movie_gross_df = movie_gross_df.drop('foreign_gross', axis = 1) # Drop the 'foreign_gross column' 
movie_gross_df

title      studio  domestic_gross  \
0                                     Toy Story 3          BV     415000000.0   
1                      Alice in Wonderland (2010)          BV     334200000.0   
2     Harry Potter and the Deathly Hallows Part 1          WB     296000000.0   
3                                       Inception          WB     292600000.0   
4                             Shrek Forever After        P/DW     238700000.0   
...                                           ...         ...             ...   
3382                                    The Quake       Magn.          6200.0   
3383                  Edward II (2018 re-release)          FM          4800.0   
3384                                     El Pacto        Sony          2500.0   
3385                                     The Swan  Synergetic          2400.0   
3386                            An Actor Prepares       Grav.          1700.0   

           year  
0    2010-01-01  
1    2010-01-01  
2    2010-01-01  
3    2010-01-01  
4    2010-01-01  
...         ...  
3382 2018-01-01  
3383 2018-01-01  
3384 2018-01-01  
3385 2018-01-01  
3386 2018-01-01  

[3387 rows x 4 columns]

In [85]:
movie_gross_df.loc[movie_gross_df['domestic_gross'].isnull() == True, : ] #Checking the entries with null values in the 'domestic_gross' column
clean_movie_gross_df = movie_gross_df.dropna(axis = 0) # Dropping all rows with null values
clean_movie_gross_df.describe().round(1)

domestic_gross
count          3356.0
mean       28771489.6
std        67006943.1
min             100.0
25%          120000.0
50%         1400000.0
75%        27950000.0
max       936700000.0

## 3
Moving on to the next file


In [145]:
dataframe_structure(movie_budgets_df)

The shape of the dataframe is: (5782, 5)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5782 entries, 1 to 82
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   release_date       5782 non-null   object
 1   movie              5782 non-null   object
 2   production_budget  5782 non-null   object
 3   domestic_gross     5782 non-null   object
 4   worldwide_gross    5782 non-null   object
dtypes: object(5)
memory usage: 271.0+ KB


release_date      movie production_budget domestic_gross  \
count           5782       5782              5782           5782   
unique          2418       5698               509           5164   
top     Dec 31, 2014  Halloween       $20,000,000             $0   
freq              24          3               231            548   

       worldwide_gross  
count             5782  
unique            5356  
top                 $0  
freq               367

Checking for Null Values: 
 release_date         0
movie                0
production_budget    0
domestic_gross       0
worldwide_gross      0
dtype: int64


release_date                                        movie  \
id                                                              
1   Dec 18, 2009                                       Avatar   
2   May 20, 2011  Pirates of the Caribbean: On Stranger Tides   
3    Jun 7, 2019                                 Dark Phoenix   
4    May 1, 2015                      Avengers: Age of Ultron   
5   Dec 15, 2017            Star Wars Ep. VIII: The Last Jedi   

   production_budget domestic_gross worldwide_gross  
id                                                   
1       $425,000,000   $760,507,625  $2,776,345,279  
2       $410,600,000   $241,063,875  $1,045,663,875  
3       $350,000,000    $42,762,350    $149,762,350  
4       $330,600,000   $459,005,868  $1,403,013,963  
5       $317,000,000   $620,181,382  $1,316,721,747

## Understanding the data
There are 5 columns and 5782 entries in `tn.movie_budgets.csv` file and `NaN` values seem to be absent .

The datatype of `release_date` should be converted to `datetime`, and the datatypes of `production_budget`, `domestic_gross`, and `worldwide_gross` should be converted to `int`.

In [182]:
movie_budgets_df['release_date'] = pd.to_datetime(movie_budgets_df['release_date']) # Converting the 'release_date' to datetime datatype

In [148]:
movie_budgets_df

release_date                                        movie  \
id                                                             
1    2009-12-18                                       Avatar   
2    2011-05-20  Pirates of the Caribbean: On Stranger Tides   
3    2019-06-07                                 Dark Phoenix   
4    2015-05-01                      Avengers: Age of Ultron   
5    2017-12-15            Star Wars Ep. VIII: The Last Jedi   
..          ...                                          ...   
78   2018-12-31                                       Red 11   
79   1999-04-02                                    Following   
80   2005-07-13                Return to the Land of Wonders   
81   2015-09-29                         A Plague So Pleasant   
82   2005-08-05                            My Date With Drew   

   production_budget domestic_gross worldwide_gross  
id                                                   
1       $425,000,000   $760,507,625  $2,776,345,279  
2       $410,600,000   $241,063,875  $1,045,663,875  
3       $350,000,000    $42,762,350    $149,762,350  
4       $330,600,000   $459,005,868  $1,403,013,963  
5       $317,000,000   $620,181,382  $1,316,721,747  
..               ...            ...             ...  
78            $7,000             $0              $0  
79            $6,000        $48,482        $240,495  
80            $5,000         $1,338          $1,338  
81            $1,400             $0              $0  
82            $1,100       $181,041        $181,041  

[5782 rows x 5 columns]

The `production_budget`, `domestic_gross`, and `worldwide_gross` columns are converted from a string values to a numeric values. Befor the conversion, the string values are first stripped off any non-numeric characters using the `replace()` function.

In [183]:
movie_budgets_df = movie_budgets_df.assign(production_budget = pd.to_numeric(movie_budgets_df['production_budget'].str.replace('[^0-9.]', '')))
movie_budgets_df = movie_budgets_df.assign(domestic_gross = pd.to_numeric(movie_budgets_df['domestic_gross'].str.replace('[^0-9.]', '')))
movie_budgets_df = movie_budgets_df.assign(worldwide_gross = pd.to_numeric(movie_budgets_df['worldwide_gross'].str.replace('[^0-9.]', '')))

release_date                                        movie  \
id                                                             
1    2009-12-18                                       Avatar   
2    2011-05-20  Pirates of the Caribbean: On Stranger Tides   
3    2019-06-07                                 Dark Phoenix   
4    2015-05-01                      Avengers: Age of Ultron   
5    2017-12-15            Star Wars Ep. VIII: The Last Jedi   
..          ...                                          ...   
78   2018-12-31                                       Red 11   
79   1999-04-02                                    Following   
80   2005-07-13                Return to the Land of Wonders   
81   2015-09-29                         A Plague So Pleasant   
82   2005-08-05                            My Date With Drew   

    production_budget  domestic_gross  worldwide_gross  
id                                                      
1           425000000       760507625       2776345279  
2           410600000       241063875       1045663875  
3           350000000        42762350        149762350  
4           330600000       459005868       1403013963  
5           317000000       620181382       1316721747  
..                ...             ...              ...  
78               7000               0                0  
79               6000           48482           240495  
80               5000            1338             1338  
81               1400               0                0  
82               1100          181041           181041  

[5782 rows x 5 columns]

## Cleaning the data

First check for duplicates. 
We find out that there are no duplicate entries in `movie_budgets_df`.

In [207]:
duplicates = movie_budgets_df[movie_budgets_df.duplicated()] # Checking for duplicate entries
len(duplicates)

0

In [216]:
movie_budgets_df.describe().round(0)

production_budget  domestic_gross  worldwide_gross
count             5782.0          5782.0     5.782000e+03
mean          31587757.0      41873327.0     9.148746e+07
std           41812077.0      68240597.0     1.747200e+08
min               1100.0             0.0     0.000000e+00
25%            5000000.0       1429534.0     4.125415e+06
50%           17000000.0      17225945.0     2.798445e+07
75%           40000000.0      52348662.0     9.764584e+07
max          425000000.0     936662225.0     2.776345e+09

In [219]:
for column in movie_budgets_df.columns: # Checking for any other data descrepancies or extraneous values
    print(column, '\n', movie_budgets_df[column].value_counts(normalize = True), '\n\n')

release_date 
 2014-12-31    0.004151
2015-12-31    0.003978
2010-12-31    0.002594
2008-12-31    0.002421
2013-12-31    0.002248
                ...   
2018-04-27    0.000173
1982-09-17    0.000173
2010-02-10    0.000173
1966-07-06    0.000173
1977-07-13    0.000173
Name: release_date, Length: 2418, dtype: float64 


movie 
 Halloween              0.000519
Home                   0.000519
King Kong              0.000519
Crossroads             0.000346
Pinocchio              0.000346
                         ...   
The Lost City of Z     0.000173
Father's Day           0.000173
The Patriot            0.000173
Konferenz der Tiere    0.000173
Der Untergang          0.000173
Name: movie, Length: 5698, dtype: float64 


production_budget 
 20000000    0.039952
10000000    0.036666
30000000    0.030612
15000000    0.029920
25000000    0.029575
              ...   
16400000    0.000173
103000      0.000173
6537890     0.000173
780000      0.000173
28500000    0.000173
Name: production_budget,


## 4th File

In [221]:
dataframe_structure(movie_info_df)

The shape of the dataframe is: (1560, 12)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1560 entries, 0 to 1559
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            1560 non-null   int64 
 1   synopsis      1498 non-null   object
 2   rating        1557 non-null   object
 3   genre         1552 non-null   object
 4   director      1361 non-null   object
 5   writer        1111 non-null   object
 6   theater_date  1201 non-null   object
 7   dvd_date      1201 non-null   object
 8   currency      340 non-null    object
 9   box_office    340 non-null    object
 10  runtime       1530 non-null   object
 11  studio        494 non-null    object
dtypes: int64(1), object(11)
memory usage: 146.4+ KB


id
count  1560.000000
mean   1007.303846
std     579.164527
min       1.000000
25%     504.750000
50%    1007.500000
75%    1503.250000
max    2000.000000

Checking for Null Values: 
 id                 0
synopsis          62
rating             3
genre              8
director         199
writer           449
theater_date     359
dvd_date         359
currency        1220
box_office      1220
runtime           30
studio          1066
dtype: int64


id                                           synopsis rating  \
0   1  This gritty, fast-paced, and innovative police...      R   
1   3  New York City, not-too-distant-future: Eric Pa...      R   
2   5  Illeana Douglas delivers a superb performance ...      R   
3   6  Michael Douglas runs afoul of a treacherous su...      R   
4   7                                                NaN     NR   

                                 genre          director  \
0  Action and Adventure|Classics|Drama  William Friedkin   
1    Drama|Science Fiction and Fantasy  David Cronenberg   
2    Drama|Musical and Performing Arts    Allison Anders   
3           Drama|Mystery and Suspense    Barry Levinson   
4                        Drama|Romance    Rodney Bennett   

                            writer  theater_date      dvd_date currency  \
0                   Ernest Tidyman   Oct 9, 1971  Sep 25, 2001      NaN   
1     David Cronenberg|Don DeLillo  Aug 17, 2012   Jan 1, 2013        $   
2                   Allison Anders  Sep 13, 1996  Apr 18, 2000      NaN   
3  Paul Attanasio|Michael Crichton   Dec 9, 1994  Aug 27, 1997      NaN   
4                     Giles Cooper           NaN           NaN      NaN   

  box_office      runtime             studio  
0        NaN  104 minutes                NaN  
1    600,000  108 minutes  Entertainment One  
2        NaN  116 minutes                NaN  
3        NaN  128 minutes                NaN  
4        NaN  200 minutes                NaN

## Understanding the data
There are 12 columns and 1560 entries in `rt.movie_info.tsv` file and all the columns have string values apart from `id` column which has integer values. All the columns apart form the `id` column have `NaN` values.

The datatype of `theater_date` and `dvd_date` should be converted to `datetime`, and the datatypes of `box_office` and `runtime` should be converted to `int` values.

## Cleaning the data

1. Check for duplicates and remove them. There are no duplicates in the `movie_info_df`.
2. Deal with null values
3. Changing column types.



In [234]:
duplicates = movie_info_df[movie_info_df.duplicated()] # Checking for duplicate entries
len(duplicates)

0

In [253]:
movie_info_df['theater_date'] = pd.to_datetime(movie_info_df['theater_date']) # Converting the 'theater_date' to datetime datatype
movie_info_df['dvd_date'] = pd.to_datetime(movie_info_df['dvd_date']) # Converting the 'dvd_date' to datetime datatype

In [254]:
movie_info_df = movie_info_df.assign(box_office = pd.to_numeric(movie_info_df['box_office'].str.replace('[^0-9.]', '')))
movie_info_df = movie_info_df.assign(runtime = pd.to_numeric(movie_info_df['runtime'].str.replace('[^0-9.]', '')))

In [255]:
movie_info_df

id                                           synopsis rating  \
0        1  This gritty, fast-paced, and innovative police...      R   
1        3  New York City, not-too-distant-future: Eric Pa...      R   
2        5  Illeana Douglas delivers a superb performance ...      R   
3        6  Michael Douglas runs afoul of a treacherous su...      R   
4        7                                                NaN     NR   
...    ...                                                ...    ...   
1555  1996  Forget terrorists or hijackers -- there's a ha...      R   
1556  1997  The popular Saturday Night Live sketch was exp...     PG   
1557  1998  Based on a novel by Richard Powell, when the l...      G   
1558  1999  The Sandlot is a coming-of-age story about a g...     PG   
1559  2000  Suspended from the force, Paris cop Hubert is ...      R   

                                                  genre            director  \
0                   Action and Adventure|Classics|Drama    William Friedkin   
1                     Drama|Science Fiction and Fantasy    David Cronenberg   
2                     Drama|Musical and Performing Arts      Allison Anders   
3                            Drama|Mystery and Suspense      Barry Levinson   
4                                         Drama|Romance      Rodney Bennett   
...                                                 ...                 ...   
1555   Action and Adventure|Horror|Mystery and Suspense                 NaN   
1556                 Comedy|Science Fiction and Fantasy        Steve Barron   
1557  Classics|Comedy|Drama|Musical and Performing Arts      Gordon Douglas   
1558    Comedy|Drama|Kids and Family|Sports and Fitness  David Mickey Evans   
1559  Action and Adventure|Art House and Internation...                 NaN   

                                                writer theater_date  \
0                                       Ernest Tidyman   1971-10-09   
1                         David Cronenberg|Don DeLillo   2012-08-17   
2                                       Allison Anders   1996-09-13   
3                      Paul Attanasio|Michael Crichton   1994-12-09   
4                                         Giles Cooper          NaT   
...                                                ...          ...   
1555                                               NaN   2006-08-18   
1556  Terry Turner|Tom Davis|Dan Aykroyd|Bonnie Turner   1993-07-23   
1557                                               NaN   1962-01-01   
1558                  David Mickey Evans|Robert Gunter   1993-04-01   
1559                                        Luc Besson   2001-09-27   

       dvd_date currency  box_office  runtime             studio  
0    2001-09-25      NaN         NaN    104.0                NaN  
1    2013-01-01        $    600000.0    108.0  Entertainment One  
2    2000-04-18      NaN         NaN    116.0                NaN  
3    1997-08-27      NaN         NaN    128.0                NaN  
4           NaT      NaN         NaN    200.0                NaN  
...         ...      ...         ...      ...                ...  
1555 2007-01-02        $  33886034.0    106.0    New Line Cinema  
1556 2001-04-17      NaN         NaN     88.0  Paramount Vantage  
1557 2004-05-11      NaN         NaN    111.0                NaN  
1558 2002-01-29      NaN         NaN    101.0                NaN  
1559 2003-02-11      NaN         NaN     94.0  Columbia Pictures  

[1560 rows x 12 columns]


## 5th File

In [256]:
dataframe_structure(reviews_df)

The shape of the dataframe is: (54432, 8)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54432 entries, 0 to 54431
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id          54432 non-null  int64 
 1   review      48869 non-null  object
 2   rating      40915 non-null  object
 3   fresh       54432 non-null  object
 4   critic      51710 non-null  object
 5   top_critic  54432 non-null  int64 
 6   publisher   54123 non-null  object
 7   date        54432 non-null  object
dtypes: int64(2), object(6)
memory usage: 3.3+ MB


id    top_critic
count  54432.000000  54432.000000
mean    1045.706882      0.240594
std      586.657046      0.427448
min        3.000000      0.000000
25%      542.000000      0.000000
50%     1083.000000      0.000000
75%     1541.000000      0.000000
max     2000.000000      1.000000

Checking for Null Values: 
 id                0
review         5563
rating        13517
fresh             0
critic         2722
top_critic        0
publisher       309
date              0
dtype: int64


id                                             review rating   fresh  \
0   3  A distinctly gallows take on contemporary fina...    3/5   fresh   
1   3  It's an allegory in search of a meaning that n...    NaN  rotten   
2   3  ... life lived in a bubble in financial dealin...    NaN   fresh   
3   3  Continuing along a line introduced in last yea...    NaN   fresh   
4   3             ... a perverse twist on neorealism...     NaN   fresh   

           critic  top_critic         publisher               date  
0      PJ Nabarro           0   Patrick Nabarro  November 10, 2018  
1  Annalee Newitz           0           io9.com       May 23, 2018  
2    Sean Axmaker           0  Stream on Demand    January 4, 2018  
3   Daniel Kasman           0              MUBI  November 16, 2017  
4             NaN           0      Cinema Scope   October 12, 2017


## Understanding the data
There are 8 columns and 54432 entries in `rt.reviews.tsv` file and all the columns have string values apart from `id` and `top_critic` which have integer values. The following columns have `NaN` values: `review`, `rating`, `critic` and `publisher`.

The datatype of values in `date` column should be converted to `datetime`

## Cleaning the data

1. Check for duplicates and remove them. There are 9 duplicate entries in `movie_info_df` are identified and removed.
2. Convert the `date` column's datatype to datetime.
3. Deal with null values
   - The `rating` column is dropped because it has 25% of its entries being null. It's data is closely linked to the `fresh`    column which has no null values.
   - The null values in `review`, `critic` and `publisher` columns will be replaced with `No Description`, `No Name` and `Missing Publisher Name` respectively. These columns and their entries have been kept because they have invaluable data.

In [279]:
duplicates = reviews_df[reviews_df.duplicated()] # Checking for duplicate entries
reviews_df_no_duplicates = reviews_df.drop_duplicates() # Drop the duplicate rows

In [264]:
reviews_df['date'] = pd.to_datetime(reviews_df['date']) # Converting the 'date' column values to datetime datatype

In [280]:
dataframe_structure(reviews_df_no_duplicates)

The shape of the dataframe is: (54423, 8)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 54423 entries, 0 to 54431
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   id          54423 non-null  int64         
 1   review      48867 non-null  object        
 2   rating      40907 non-null  object        
 3   fresh       54423 non-null  object        
 4   critic      51710 non-null  object        
 5   top_critic  54423 non-null  int64         
 6   publisher   54114 non-null  object        
 7   date        54423 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(2), object(5)
memory usage: 3.7+ MB


id    top_critic
count  54423.000000  54423.000000
mean    1045.668596      0.240634
std      586.661865      0.427472
min        3.000000      0.000000
25%      542.000000      0.000000
50%     1083.000000      0.000000
75%     1541.000000      0.000000
max     2000.000000      1.000000

Checking for Null Values: 
 id                0
review         5556
rating        13516
fresh             0
critic         2713
top_critic        0
publisher       309
date              0
dtype: int64


id                                             review rating   fresh  \
0   3  A distinctly gallows take on contemporary fina...    3/5   fresh   
1   3  It's an allegory in search of a meaning that n...    NaN  rotten   
2   3  ... life lived in a bubble in financial dealin...    NaN   fresh   
3   3  Continuing along a line introduced in last yea...    NaN   fresh   
4   3             ... a perverse twist on neorealism...     NaN   fresh   

           critic  top_critic         publisher       date  
0      PJ Nabarro           0   Patrick Nabarro 2018-11-10  
1  Annalee Newitz           0           io9.com 2018-05-23  
2    Sean Axmaker           0  Stream on Demand 2018-01-04  
3   Daniel Kasman           0              MUBI 2017-11-16  
4             NaN           0      Cinema Scope 2017-10-12

In [294]:
reviews_df_no_duplicates.drop('rating', axis = 1, inplace = True) # Drop the 'rating' column 

In [295]:
reviews_df_no_duplicates

id                                             review   fresh  \
0         3  A distinctly gallows take on contemporary fina...   fresh   
1         3  It's an allegory in search of a meaning that n...  rotten   
2         3  ... life lived in a bubble in financial dealin...   fresh   
3         3  Continuing along a line introduced in last yea...   fresh   
4         3             ... a perverse twist on neorealism...    fresh   
...     ...                                                ...     ...   
54427  2000  The real charm of this trifle is the deadpan c...   fresh   
54428  2000                                                NaN  rotten   
54429  2000                                                NaN  rotten   
54430  2000                                                NaN  rotten   
54431  2000                                                NaN   fresh   

                  critic  top_critic         publisher       date  
0             PJ Nabarro           0   Patrick Nabarro 2018-11-10  
1         Annalee Newitz           0           io9.com 2018-05-23  
2           Sean Axmaker           0  Stream on Demand 2018-01-04  
3          Daniel Kasman           0              MUBI 2017-11-16  
4                    NaN           0      Cinema Scope 2017-10-12  
...                  ...         ...               ...        ...  
54427      Laura Sinagra           1     Village Voice 2002-09-24  
54428  Michael Szymanski           0        Zap2it.com 2005-09-21  
54429       Emanuel Levy           0   EmanuelLevy.Com 2005-07-17  
54430   Christopher Null           0    Filmcritic.com 2003-09-07  
54431    Nicolas Lacroix           0      Showbizz.net 2002-11-12  

[54423 rows x 7 columns]

In [302]:
reviews_df_no_duplicates.loc[reviews_df_no_duplicates['review'].isnull() == True, 'review'] = 'No Description'
reviews_df_no_duplicates.loc[reviews_df_no_duplicates['critic'].isnull() == True, 'critic'] = 'No Name'
reviews_df_no_duplicates.loc[reviews_df_no_duplicates['publisher'].isnull() == True, 'publisher'] = 'No Publisher Name'

In [303]:
reviews_df_no_duplicates.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 54423 entries, 0 to 54431
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   id          54423 non-null  int64         
 1   review      54423 non-null  object        
 2   fresh       54423 non-null  object        
 3   critic      54423 non-null  object        
 4   top_critic  54423 non-null  int64         
 5   publisher   54423 non-null  object        
 6   date        54423 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(2), object(4)
memory usage: 3.3+ MB


In [304]:
reviews_df_no_duplicates

id                                             review   fresh  \
0         3  A distinctly gallows take on contemporary fina...   fresh   
1         3  It's an allegory in search of a meaning that n...  rotten   
2         3  ... life lived in a bubble in financial dealin...   fresh   
3         3  Continuing along a line introduced in last yea...   fresh   
4         3                                            No Name   fresh   
...     ...                                                ...     ...   
54427  2000  The real charm of this trifle is the deadpan c...   fresh   
54428  2000                                     No Description  rotten   
54429  2000                                     No Description  rotten   
54430  2000                                     No Description  rotten   
54431  2000                                     No Description   fresh   

                  critic  top_critic         publisher       date  
0             PJ Nabarro           0   Patrick Nabarro 2018-11-10  
1         Annalee Newitz           0           io9.com 2018-05-23  
2           Sean Axmaker           0  Stream on Demand 2018-01-04  
3          Daniel Kasman           0              MUBI 2017-11-16  
4                No Name           0      Cinema Scope 2017-10-12  
...                  ...         ...               ...        ...  
54427      Laura Sinagra           1     Village Voice 2002-09-24  
54428  Michael Szymanski           0        Zap2it.com 2005-09-21  
54429       Emanuel Levy           0   EmanuelLevy.Com 2005-07-17  
54430   Christopher Null           0    Filmcritic.com 2003-09-07  
54431    Nicolas Lacroix           0      Showbizz.net 2002-11-12  

[54423 rows x 7 columns]